ReactAndroid/src/main/java/com/facebook/react/modules/settings/SettingsModule.java

```diff
@@ -0,0 +1,180 @@
+  static final String TAG = "SettingsMmodule";
+
+  private static boolean sSettingsModuleInstantiated = false;
+
+  static private String sFilename = "";
+  private Boolean ignoringUpdates = false;
+  private SharedPreferences mPreferences;
+
+  public SettingsModule(ReactApplicationContext context) {
+    super(context);
+
+  }
+
+  @Override
+  public String getName() {
+    return SettingsModule.NAME;
+  }
+
+  private SharedPreferences ensurePreferences() {
+    if (mPreferences == null) {
+      if (SettingsModule.sFilename.length() > 0) {
+        mPreferences = getReactApplicationContext().getSharedPreferences(SettingsModule.sFilename, Context.MODE_PRIVATE);
+      } else if (getCurrentActivity() != null) {
+        mPreferences = getCurrentActivity().getPreferences(Context.MODE_PRIVATE);
+      } else {
+        mPreferences = getReactApplicationContext().getSharedPreferences(DEFAULT_FILE_NAME, Context.MODE_PRIVATE);
+      }
+      sSettingsModuleInstantiated = true;
+      mPreferences.registerOnSharedPreferenceChangeListener(this);
+    }
+
+    return mPreferences;
+  }
+
+  /***
+   * Configure the Settings Module to use a SharedPreferences file with a specific name as its backing store.
+   * Useful if the Settings Module needs to use an existing SharedPreferences file or to ensure that Settings Modules
+   * belonging to different ReactPackages will all share the same SharedPreferences file for their backing store.
+   * <p></p>
+   * <b>Note:</b>this method may only be called before a Settings Module initialization (i.e. before a ReactMainPackage is instantiated)
+   * An app would typically invoke this function in the app's main activity before setting up the React Native environment
+   * @param filename the name of the SharedPreferences file to use or create for the backing store
+   * @throws IllegalStateException if called after instantiation of a Settings module
+   */
+  static public void setSharedPreferencesFilename(String filename) {
+    if (sSettingsModuleInstantiated) {
+      throw new IllegalStateException("Attempt to set Settings Module SharedPreferences file name after a Settings Module instantiation");
+    }
+    sFilename = filename;
+  }
+
+  @Override
+  public void onCatalystInstanceDestroy() {
+    ensurePreferences().unregisterOnSharedPreferenceChangeListener(this);
+  }
+
+  @ReactMethod
+  public void setValues(ReadableMap values) {
+    this.ignoringUpdates = true;
+
+    ReadableMapKeySetIterator iterator = values.keySetIterator();
+    SharedPreferences.Editor editor = mPreferences.edit();
+
+    while (iterator.hasNextKey()) {
+      String key = iterator.nextKey();
+      ReadableType type = values.getType(key);
+
+      switch (type) {
+        case Null:
+          editor.remove(key);
+          break;
+        case Boolean:
+          editor.putBoolean(key, values.getBoolean(key));
+          break;
+        case Number:
+          editor.putFloat(key, (float) values.getDouble(key));
```

Comment:
Remember these are the _persisted_ settings values, not the _current_ values of the Settings Module. Whereas AsyncStorage has to read and write its stored values asynchronously through native code, Settings reads and writes its settings immediately and synchronously all within its javascript code. It is true that the JS-side settings value are then lazily and asynchronously propagated to native code - both to inform the native app of any changes and also to persist the settings values for a subsequent relaunch of the app. But the values being stored here are **not** the values that Settings Module is actually using.

Therefore, there is no reason to use a promise here because the Settings module is basically unaware and unaffected by this method is doing.

The double-to-float conversion does affects the persisted numeric values that populate the Settings Module upon launch. Because these loaded numeric values have been "round tripped" through a smaller type, they may wind up losing precision or over/under flowing their representation. Although this information loss is unfortunate - as a practical matter, few settings are likely to be affected - either because they are integers or they are far, far smaller than 3.4 × 10³⁸ - the upper range of a float.

Moreover, it is not a error for a number to be too large or too small a floating point type to represent - for such numbers, infinity is used for the huge ones and zero for the tiny values.